// vite-plugin-ssr/dist/cjs/node/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.__private = exports.RenderErrorPage = exports._injectAssets = exports.pipeNodeStream = exports.pipeWebStream = exports.dangerouslySkipEscape = exports.escapeInject = exports.createPageRenderer = void 0;
var createPageRenderer_1 = require("./createPageRenderer");
Object.defineProperty(exports, "createPageRenderer", { enumerable: true, get: function () { return createPageRenderer_1.createPageRenderer; } });
var renderHtml_1 = require("./html/renderHtml");
Object.defineProperty(exports, "escapeInject", { enumerable: true, get: function () { return renderHtml_1.escapeInject; } });
Object.defineProperty(exports, "dangerouslySkipEscape", { enumerable: true, get: function () { return renderHtml_1.dangerouslySkipEscape; } });
var stream_1 = require("./html/stream");
Object.defineProperty(exports, "pipeWebStream", { enumerable: true, get: function () { return stream_1.pipeWebStream; } });
Object.defineProperty(exports, "pipeNodeStream", { enumerable: true, get: function () { return stream_1.pipeNodeStream; } });
var injectAssets_1 = require("./html/injectAssets");
Object.defineProperty(exports, "_injectAssets", { enumerable: true, get: function () { return injectAssets_1.injectAssets__public; } });
var RenderErrorPage_1 = require("./renderPage/RenderErrorPage");
Object.defineProperty(exports, "RenderErrorPage", { enumerable: true, get: function () { return RenderErrorPage_1.RenderErrorPage; } });
require("./page-files/setup");
const importBuild_1 = require("./importBuild");
const utils_1 = require("./utils");
exports.__private = { importBuild: importBuild_1.importBuild };
(0, utils_1.assertUsage)(!(0, utils_1.isBrowser)(), 'The `vite-plugin-ssr` module cannot be imported in the browser.');
